scalafix-core/shared/src/main/scala/scalafix/internal/rule/OrphanImplicits.scala

```diff
@@ -0,0 +1,85 @@
+package scalafix.internal.rule
+
+import scala.meta._
+import scalafix.internal.util.{SymbolOps, TreeOps}
+import scalafix.rule.RuleCtx
+import scalafix.syntax._
+import scalafix.{LintCategory, LintMessage, SemanticRule, SemanticdbIndex}
+
+final case class OrphanImplicits(index: SemanticdbIndex)
+    extends SemanticRule(index, "OrphanImplicits") {
+
+  override def description: String =
+    "Linter that reports an error on implicit instances of shape F[G] " +
+      "that do not belong to the companion object of F or G."
+
+  private lazy val errorCategory: LintCategory =
+    LintCategory.error("")
+
+  override def check(ctx: RuleCtx): Seq[LintMessage] = {
+
+    def handleImplicit(
+        tpe: Type,
+        obj: Symbol.Global,
+        pos: Position): Option[LintMessage] = {
+
+      val tpeSymbols = tpe match {
+        case Type.Apply(t, Seq(arg)) =>
+          Seq(index.symbol(t), index.symbol(arg)).flatten
+        case _ => Seq.empty
+      }
+
+      if (tpeSymbols.nonEmpty && !tpeSymbols.exists(
+          SymbolOps.isSameNormalized(obj, _)
+        )) {
+        Some(
+          errorCategory
+            .at(
+              s"""Orphan implicits are not allowed.
+                 |This definition is only allowed in one of the following objects:
+                 |${tpeSymbols.mkString(", ")}
+                """.stripMargin,
+              pos
+            ))
+      } else {
+        None
+      }
+    }
+    def handleDefn(
+        mods: List[Mod],
+        tpe: Option[Type],
+        d: Defn,
+        obj: Symbol.Global): Option[LintMessage] = {
+      if (mods.exists(_.is[Mod.Implicit])) {
+        tpe match {
+          case Some(t) => handleImplicit(t, obj, d.pos)
```

Comment:
This would report false positives on situations like
```scala
trait Writer[T]
object Writer {
  type IntWriter = Writer[Int]
  implicit val x: IntWriter = ...
}
```
I would not rely on the syntactic structure of the type in the source.

Reply:
Ahh, you're right! The new semanticdb Type does handle that case however, you can leave it unchanged for now 😉